### PR TITLE
Fixes #26753 - react-router foundation #3

### PR DESCRIPTION
**Pull request by glekner**
add `onClick` prop to Bookmarks to disable`Turbolinks.visit` when not needed.

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js [hunk 1]
```diff
@@ -1,8 +1,7 @@
-import URI from 'urijs';
+import { getURI, changeQuery } from '../Pagination/PaginationHelper';
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js]: * Isn't it odd you are importing `getURI` and `changeQuery` from `PaginationHelper`?
* isn't it suppose to use `react-router` now for those type of operations?

Comment (glekner) [reply webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js]: this function is for the current views not for the `react-router` ones. either way I think its best to move all URI functions to a dedicated helper.

**Comment by glekner** [comment]
- Moved all URI helpers to dedicated file.

**Comment by sharvit** [comment]
> * Moved all URI helpers to dedicated file.

@glekner notice @MariaAga commit that creates the same helper with code moved from `katello`:
https://github.com/theforeman/foreman/commit/6730cf2a5b440021373a5687dcc2adacfd61bd1d#diff-5583b0c21a9139208c8282c8618869a2

**Comment by glekner** [comment]
Summarizing changes in this PR:

- moved all URI funcs/helpers to `urlHelpers`
- added Bookmarks functionality for `react-router`
- added missing css classes for collapse nav state
- added empty react controller view and added it to permissions
- PageLayout now changes `document.title` on render

File: app/registries/foreman/access_permissions.rb [hunk 2]
```diff
@@ -756,4 +756,8 @@
     map.permission :edit_settings, { :settings => [:update],
                                      :'api/v2/settings' => [:update] }
   end
+
+  permission_set.security_block :react do |map|
+    map.permission :view_react, {:react => [:index]}
```
Comment (sharvit) [inline app/registries/foreman/access_permissions.rb:761]: I'm confused, why is this needed?
Do we need special permission to see react pages or it should just be allowed to everyone?

File: webpack/assets/javascripts/react_app/common/urlHelpers.js [hunk 3]
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
...
+
+/**
+ * Get updated URI
+ */
+export const getURI = () => new URI(window.location.href);
+
+/**
+ * Get updated page param
+ */
+export const getURIpage = () => Number(getURI().query(true).page) || 1;
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.js:27]: test?

File: webpack/assets/javascripts/react_app/common/urlHelpers.js [hunk 4]
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
...
+/**
+ * Get updated perPage param
+ */
+export const getURIperPage = () => Number(getURI().query(true).per_page);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.js:31]: test?

File: webpack/assets/javascripts/react_app/common/urlHelpers.js [hunk 5]
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
...
+/**
+ * Get updated searchQuery param
+ */
+export const getURIsearch = () => getURI().query(true).search || '';
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.js:35]: test?

File: webpack/assets/javascripts/react_app/common/urlHelpers.js [hunk 6]
```diff
@@ -13,3 +15,60 @@ export const urlBuilder = (controller, action, id = undefined) =>
...
+
+/**
+ * Get updated URI params
+ */
+export const getParams = () => ({
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.js:40]: test?

File: webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js [hunk 7]
```diff
@@ -1,8 +1,7 @@
-import URI from 'urijs';
+import { getURI, changeQuery } from '../../common/urlHelpers';
 
 export const resolveSearchQuery = searchQuery => {
-  const uri = new URI(window.location.href);
+  const uri = getURI();
   const data = { ...uri.query(true), search: searchQuery.trim(), page: 1 };
-  uri.query(URI.buildQuery(data, true));
-  window.Turbolinks.visit(uri.toString());
+  changeQuery(uri, data);
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js]: Do we have to call `getURI()` and pass it to the `changeQuery` every time we want to use it?
Would it make sense if `changeQuery` will call `getURI` instead of the consumer?

File: webpack/assets/javascripts/react_app/pages/common/PageLayout/PageLayout.js [hunk 8]
```diff
@@ -13,54 +13,59 @@ const PageLayout = ({
...
-        >
-          <ToastsList />
-        </div>
-      )}
-      <div id="breadcrumb">
-        {!breadcrumbOptions && (
-          <div className="row form-group">
-            <h1 className="col-md-8">{header}</h1>
+}) => {
+  document.title = header;
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/pages/common/PageLayout/PageLayout.js]: Can we create a small helper function for this operation (e.g. `updateDocumentTitle(header)`)?
IMO it can live here: https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/common/document.js

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @glekner

File: app/registries/foreman/access_permissions.rb (hunk 2 above)
Comment (glekner) [reply app/registries/foreman/access_permissions.rb:761]: these changes are needed for a rake test to pass (permissions_test, seed_test)

File: webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js (hunk 7 above)
Comment (glekner) [reply webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js]: Sometime consumers need to `getURI()` before calling `changeQuery` (manipulating data etc), so if we put it inside it will be called twice for no reason.

File: webpack/assets/javascripts/react_app/common/__snapshots__/urlHelpers.test.js.snap [hunk 9]
```diff
@@ -0,0 +1,17 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`URI query and stringify tests should return stringified params 1`] = `"?page=3&per_page=21&search=search"`;
+
+exports[`URI query and stringify tests should test params functions 1`] = `
+Object {
+  "page": 1,
+  "perPage": NaN,
+  "searchQuery": "",
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/__snapshots__/urlHelpers.test.js.snap]: Can we simulate real values for the tests?
Maybe by updating the `window.location`?

File: webpack/assets/javascripts/react_app/common/urlHelpers.test.js [hunk 10]
```diff
@@ -21,3 +31,34 @@ describe('urlWithSearch', () => {
...
+    const params = { page: 3, perPage: 21, searchQuery: 'search' };
+    const stringified = stringifyParams(params);
+    expect(stringified).toMatchSnapshot();
+  });
+
+  it('should test params functions', () => {
+    expect(getParams()).toMatchSnapshot();
+    expect(getURIpage()).toMatchSnapshot();
+    expect(getURIperPage()).toMatchSnapshot();
+    expect(getURIsearch()).toMatchSnapshot();
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.test.js]: Could you give a description to your snapshots? `.toMatchSnapshot('getParams')`

File: webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js (hunk 7 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/SearchBar/SearchBarHelpers.js]: What if `changeQuery` would be `changeQuery(data, uri = getURI())`?

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @glekner

**Comment by glekner** [comment]
added tests, removed `SearchBarHelper` and `PaginationHelper`
@sharvit can you ack?

**Review by sharvit — APPROVED** [review]
ACK

**Comment by glekner** [comment]
can we merge @mmoll @ohadlevy ? :)

**Review by mmoll — APPROVED** [review]

**Comment by mmoll** [comment]
merged, thanks @glekner!

File: app/registries/foreman/access_permissions.rb (hunk 2 above)
Comment (tbrisker) [reply app/registries/foreman/access_permissions.rb:761]: They shouldn't be. We don't add permissions just so tests pass when they don't make sense from a user standpoint. opened https://projects.theforeman.org/issues/27333 to remove this.
